cdc/puller/ddl_puller.go

```diff
@@ -43,6 +50,165 @@ const (
 	DDLPullerTableName = "DDL_PULLER"
 )
 
+// DDLJobPuller is used to pull ddl job from TiKV.
+type DDLJobPuller interface {
+	// Run starts the DDLJobPuller.
+	Run(ctx context.Context) error
+	// Output the DDL job entry, it contains the DDL job and the error.
+	Output() <-chan *model.DDLJobEntry
+}
+
+type ddlJobPullerImpl struct {
+	puller    Puller
+	kvStorage tidbkv.Storage
+	// tableInfo is initialized when receive the first concurrent DDL job.
+	tableInfo *model.TableInfo
+	columnID  int64
+	outputCh  chan *model.DDLJobEntry
+}
+
+// Run starts the DDLJobPuller.
+func (p *ddlJobPullerImpl) Run(ctx context.Context) error {
+	g, ctx := errgroup.WithContext(ctx)
+
+	g.Go(func() error {
+		return errors.Trace(p.puller.Run(ctx))
+	})
+
+	rawDDLCh := memory.SortOutput(ctx, p.puller.Output())
+	for {
+		var ddlRawKV *model.RawKVEntry
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case ddlRawKV = <-rawDDLCh:
+		}
+		if ddlRawKV == nil {
+			continue
+		}
+
+		job, err := p.unmarshalDDL(ddlRawKV)
+		jobEntry := &model.DDLJobEntry{
+			Job:    job,
+			OpType: ddlRawKV.OpType,
+			CRTs:   ddlRawKV.CRTs,
+			Err:    err,
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case p.outputCh <- jobEntry:
+		}
+	}
+}
+
+// Output the DDL job entry, it contains the DDL job and the error.
+func (p *ddlJobPullerImpl) Output() <-chan *model.DDLJobEntry {
+	return p.outputCh
+}
+
+func (p *ddlJobPullerImpl) initJobTableMeta() error {
+	version, err := p.kvStorage.CurrentVersion(tidbkv.GlobalTxnScope)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	snap, err := kv.GetSnapshotMeta(p.kvStorage, version.Ver)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	dbInfos, err := snap.ListDatabases()
+	if err != nil {
+		return cerror.WrapError(cerror.ErrMetaListDatabases, err)
+	}
+
+	db, err := findDBByName(dbInfos, mysql.SystemDB)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	tbls, err := snap.ListTables(db.ID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	tableInfo, err := findTableByName(tbls, "tidb_ddl_job")
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	col, err := findColumnByName(tableInfo.Columns, "job_meta")
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	p.tableInfo = model.WrapTableInfo(db.ID, db.Name.L, 0, tableInfo)
+	p.columnID = col.ID
+	return nil
+}
+
+func (p *ddlJobPullerImpl) unmarshalDDL(rawKV *model.RawKVEntry) (*timodel.Job, error) {
+	if rawKV.OpType != model.OpTypePut {
+		return nil, nil
+	}
+	if p.tableInfo == nil && !entry.IsLegacyFormatJob(rawKV) {
+		err := p.initJobTableMeta()
```

Comment:
Why lazy initialize? What if the table changes in future TiDB?

Reply:
just need the column `job_meta` if will be fine